Step Number Missing from TestCheckFunc Failures (Redux)

### SDK version

```
v2.0.1
```

### Relevant provider source code

```go
func TestAccAWSKmsExternalKey_Description(t *testing.T) {
	var key1, key2 kms.KeyMetadata
	resourceName := "aws_kms_external_key.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSKmsExternalKeyDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSKmsExternalKeyConfigDescription("description1"),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSKmsExternalKeyExists(resourceName, &key1),
					resource.TestCheckResourceAttr(resourceName, "description", "description1"),
				),
			},
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
				ImportStateVerifyIgnore: []string{
					"deletion_window_in_days",
					"key_material_base64",
				},
			},
			{
				Config: testAccAWSKmsExternalKeyConfigDescription("description2"),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSKmsExternalKeyExists(resourceName, &key2),
					testAccCheckAWSKmsExternalKeyNotRecreated(&key1, &key2),
					resource.TestCheckResourceAttr(resourceName, "description", "description2"),
				),
			},
		},
	})
}

func testAccAWSKmsExternalKeyConfigDescription(description string) string {
	return fmt.Sprintf(`
resource "aws_kms_external_key" "test" {
  description             = %[1]q
  deletion_window_in_days = 7
}
`, description)
}
```

### Terraform Configuration Files

See above. 👍 

### Debug Output

Please ask if you need it.

### Expected Behavior

Step number included in test failure message.

### Actual Behavior

A few examples from last night's full run (including the original #520 report):

```
TestAccAwsBackupPlan_withLifecycle: resource_aws_backup_plan_test.go:191: Check 7/8 error: aws_backup_plan.test: Attribute 'rule.0.lifecycle.0.delete_after' not found
--- FAIL: TestAccAwsBackupPlan_withLifecycle (10.92s)

TestAccAWSIAMPolicy_basic: resource_aws_iam_policy_test.go:102: Check 6/6 error: aws_iam_policy.test: Attribute 'policy' expected "{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"ec2:Describe*\"],\"Effect\":\"Allow\",\"Resource\":\"*\"}]}", got "{\n  \"Version\": \"2012-10-17\",\n  \"Statement\": [\n    {\n      \"Action\": [\n        \"ec2:Describe*\"\n      ],\n      \"Effect\": \"Allow\",\n      \"Resource\": \"*\"\n    }\n  ]\n}\n"
--- FAIL: TestAccAWSIAMPolicy_basic (9.09s)

TestAccAWSKmsExternalKey_Description: resource_aws_kms_external_key_test.go:117: Check 3/3 error: aws_kms_external_key.test: Attribute 'description' expected "description2", got "description1"
--- FAIL: TestAccAWSKmsExternalKey_Description (45.25s)

TestAccDataSourceAwsVpcDhcpOptions_Filter: data_source_aws_vpc_dhcp_options_test.go:52: Check 1/13 error: Not found: aws_vpc_dhcp_options.test in [root]
--- FAIL: TestAccDataSourceAwsVpcDhcpOptions_Filter (23.70s)
```

### Steps to Reproduce

1. `TF_ACC=1 go test ./aws -v -count 1 -timeout 120m -parallel 20 -run='TestAccAWSKmsExternalKey_Description'`

### References

- Similar to https://github.com/hashicorp/terraform-plugin-sdk/issues/520

--- a/helper/resource/testing_new_config.go
+++ b/helper/resource/testing_new_config.go
@@ -1,6 +1,8 @@
 package resource
 
 import (
+	"fmt"
+
 	tfjson "github.com/hashicorp/terraform-json"
 	tftest "github.com/hashicorp/terraform-plugin-test/v2"
 	testing "github.com/mitchellh/go-testing-interface"
@@ -24,7 +26,7 @@ func testStepNewConfig(t testing.T, c TestCase, wd *tftest.WorkingDir, step Test
 			return err
 		}
 		if err := testStepTaint(state, step); err != nil {
-			t.Fatalf("Error when tainting resources: %s", err)
+			return fmt.Errorf("Error when tainting resources: %s", err)
 		}
 	}
 
@@ -58,7 +60,7 @@ func testStepNewConfig(t testing.T, c TestCase, wd *tftest.WorkingDir, step Test
 		if step.Check != nil {
 			state.IsBinaryDrivenTest = true
 			if err := step.Check(state); err != nil {
-				t.Fatal(err)
+				return err
 			}
 		}
 	}
@@ -91,7 +93,7 @@ func testStepNewConfig(t testing.T, c TestCase, wd *tftest.WorkingDir, step Test
 		if err != nil {
 			return err
 		}
-		t.Fatalf("After applying this test step, the plan was not empty.\nstdout:\n\n%s", stdout)
+		return fmt.Errorf("After applying this test step, the plan was not empty.\nstdout:\n\n%s", stdout)
 	}
 
 	// do a refresh
@@ -130,7 +132,7 @@ func testStepNewConfig(t testing.T, c TestCase, wd *tftest.WorkingDir, step Test
 		if err != nil {
 			return err
 		}
-		t.Fatalf("After applying this test step and performing a `terraform refresh`, the plan was not empty.\nstdout\n\n%s", stdout)
+		return fmt.Errorf("After applying this test step and performing a `terraform refresh`, the plan was not empty.\nstdout\n\n%s", stdout)
 	}
 
 	// ID-ONLY REFRESH
@@ -164,7 +166,7 @@ func testStepNewConfig(t testing.T, c TestCase, wd *tftest.WorkingDir, step Test
 		// caught a different bug.
 		if idRefreshCheck != nil {
 			if err := testIDRefresh(c, t, wd, step, idRefreshCheck); err != nil {
-				t.Fatalf(
+				return fmt.Errorf(
 					"[ERROR] Test: ID-only test failed: %s", err)
 			}
 		}

